zerostash/src/commands/mount.rs

```diff
@@ -0,0 +1,381 @@
+            Some((a, b)) => (a, b),
+            None => return Err(ChunkDataError::NullChunkPointer),
+        };
+        let next_start = get_next_chunk_offset(file_size, chunks, self.chunk_index);
+        let mut temp_buf = vec![0; next_start - (*start as usize)];
+        objectreader.read_chunk(chunk_p, &mut temp_buf).unwrap();
+        self.stack.append(&mut temp_buf);
+        self.increment_index();
+        Ok(())
+    }
+}
+
+fn get_next_chunk_offset(
+    file_size: usize,
+    chunks: &[(u64, Arc<ChunkPointer>)],
+    chunk_index: usize,
+) -> usize {
+    let arc = (file_size as u64, Arc::new(ChunkPointer::default()));
+    let (chunk_offset, _) = chunks.get(chunk_index + 1).unwrap_or(&arc);
+    *chunk_offset as usize
+}
+
+#[derive(Debug)]
+pub enum ChunkDataError {
+    NullChunkPointer,
+}
+
+pub struct ZerostashFS {
+    pub commit_timestamp: SystemTime,
+    pub destroy_tx: mpsc::SyncSender<()>,
+    pub stash: Infinitree<Files>,
+    pub stack: Mutex<HashMap<PathBuf, ChunkDataStack>>,
+}
+
+const DIR_ATTR: FileAttr = FileAttr {
+    size: 0,
+    blocks: 0,
+    atime: SystemTime::UNIX_EPOCH,
+    mtime: SystemTime::UNIX_EPOCH,
+    ctime: SystemTime::UNIX_EPOCH,
+    crtime: SystemTime::UNIX_EPOCH,
+    kind: FileType::Directory,
+    perm: 0o444,
+    nlink: 1,
+    uid: 1000,
+    gid: 1000,
+    rdev: 0,
+    flags: 0,
+};
+
+pub fn match_filetype(file_type: &zerostash_files::FileType) -> FileType {
+    match file_type {
+        zerostash_files::FileType::File => FileType::RegularFile,
+        zerostash_files::FileType::Symlink(_) => FileType::Symlink,
+        zerostash_files::FileType::Directory => panic!("Didnt expect a directory!"),
+    }
+}
+
+pub fn walk_dir_up(index: &VersionedMap<PathBuf, Mutex<Vec<Dir>>>, path: PathBuf) {
+    if let Some(parent) = path.parent() {
+        let dir = Dir::new(path.clone(), zerostash_files::FileType::Directory);
+        match index.get(parent) {
+            Some(parent_map) => {
+                if !parent_map.lock().unwrap().contains(&dir) {
+                    parent_map.lock().unwrap().push(dir);
+                }
+            }
+            None => {
+                index.insert(parent.to_path_buf(), Mutex::new(vec![dir]));
+            }
+        }
+        walk_dir_up(index, parent.to_path_buf());
+    }
+}
+
+impl ZerostashFS {
+    pub fn open(
+        stash: Infinitree<Files>,
+        _options: &restore::Options,
+        destroy_tx: mpsc::SyncSender<()>,
+    ) -> Result<Self> {
+        stash.load_all().unwrap();
+
+        let commit_timestamp = stash.commit_list().last().unwrap().metadata.time;
+        let mut temp_paths: Vec<PathBuf> = vec![];
+
+        {
+            stash.index().directories.for_each(|k, _| {
+                temp_paths.push(k.to_path_buf());
+            });
+            for k in temp_paths.iter() {
+                let index = &stash.index().directories;
+                walk_dir_up(index, k.to_path_buf());
+            }
+        }
```

Comment:
Why do you do this? I don't understand what's happening here.

Reply:
I still need to completely change how that works but I'm basically just iterating up the dirs until I reach the root directory as its the starting point of the file system. I cant directly add the new dirs to `directories` because I'm iterating over them, so I'm temp saving the keys in 'temp_paths'.